src/sync/generic_network_transport.hpp

```diff
@@ -0,0 +1,335 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+
+#ifndef REALM_GENERIC_NETWORK_TRANSPORT_HPP
+#define REALM_GENERIC_NETWORK_TRANSPORT_HPP
+
+#include <functional>
+#include <string>
+#include <memory>
+#include <map>
+#include <vector>
+
+#include <realm/util/to_string.hpp>
+
+namespace realm {
+namespace app {
+namespace error {
+
+#pragma mark Errors
+
+struct AppError : public std::runtime_error
+{
+    enum class Type {
+        Generic,
+        Client,
+        Session,
+    };
+
+    AppError()
+    : std::runtime_error("AppError")
+    , type(Type::Generic)
+    {
+    };
+    AppError(std::string msg, Type classification = Type::Generic)
+    : std::runtime_error(msg)
+    , type(classification)
+    {
+    }
+    const Type type;
+};
+
+struct ClientError : public AppError
+{
+    enum class Code {
+        bad_token,
+        bad_response
+    };
+
+    ClientError(ClientError::Code c)
+    : AppError(error_string_for_code(c), AppError::Type::Client)
+    , code(c)
+    {
+    }
+
+    const Code code;
+
+private:
+    static const std::string error_string_for_code(const ClientError::Code& code)
+    {
+        switch (code)
+        {
+            case ClientError::Code::bad_token:
+                return "Bad Token";
+            case ClientError::Code::bad_response:
+                return "Bad Response";
+        }
+        return util::format("unknown error code: %1", static_cast<int>(code));
+    }
+};
+
+enum class ServiceErrorCode {
```

Comment:
Would be nice with some docs explaining where these error codes are defined? I assume they come from Stitch somewhere?

Reply:
They are defined in the Stitch go codebase 😅

Reply:
Can we have a link to them here for easy reference then? Even if it links to a private repo it is rather nice instead of trying to find them manually?